controllers/application_poller_controller.go

```diff
@@ -0,0 +1,308 @@
+// in the cluster.
+// note, rbac defined in cli/internal/commands/grant_permissions/generator
+func (p *Poller) handleActivity(ctx context.Context, activity applications.ActivityItem) {
+	// We always want to delete the activity after having received it
+	defer func() {
+		if err := p.apiClient.DeleteActivity(ctx, activity.URL); err != nil {
+			p.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to delete activity", err), activity.Tags[0], activity.URL)
+		}
+	}()
+
+	// Ensure we actually have an action to perform
+	if len(activity.Tags) != 1 {
+		p.handleErrors(ctx, fmt.Errorf("%s %d", "invalid number of activity tags, expected 1 got", len(activity.Tags)), "", activity.URL)
+		return
+	}
+
+	p.log.Info("starting activity task", "task", activity.Tags[0], "activity", activity.URL)
+
+	// Activity feed provides us with a scenario URL
+	scenario, err := p.apiClient.GetScenario(ctx, activity.ExternalURL)
+	if err != nil {
+		p.handleErrors(ctx, fmt.Errorf("%s (%s): %w", "failed to get scenario", activity.ExternalURL, err), activity.Tags[0], activity.URL)
+		return
+	}
+
+	// Need to fetch top level application so we can get the resources
+	applicationURL := scenario.Link(api.RelationUp)
+	if applicationURL == "" {
+		p.handleErrors(ctx, fmt.Errorf("no matching application URL for scenario"), activity.Tags[0], activity.URL)
+		return
+	}
+
+	templateURL := scenario.Link(api.RelationTemplate)
+	if templateURL == "" {
+		p.handleErrors(ctx, fmt.Errorf("no matching template URL for scenario"), activity.Tags[0], activity.URL)
+		return
+	}
+
+	experimentURL := scenario.Link(api.RelationExperiments)
+	if experimentURL == "" {
+		p.handleErrors(ctx, fmt.Errorf("no matching experiment URL for scenario"), activity.Tags[0], activity.URL)
+		return
+	}
+
+	apiApp, err := p.apiClient.GetApplication(ctx, applicationURL)
+	if err != nil {
+		p.handleErrors(ctx, fmt.Errorf("%s (%s): %w", "failed to get application", activity.URL, err), activity.Tags[0], activity.URL)
+		return
+	}
+
+	var assembledApp *optimizeappsv1alpha1.Application
+	if assembledApp, err = server.APIApplicationToClusterApplication(apiApp, scenario); err != nil {
+		p.handleErrors(ctx, fmt.Errorf("%s (%s): %w", "failed to assemble application", activity.URL, err), activity.Tags[0], activity.URL)
+		return
+	}
+
+	generatedResources, err := p.generateApp(*assembledApp)
+	if err != nil {
+		p.handleErrors(ctx, fmt.Errorf("%s (%s): %w", "failed to generate application", activity.URL, err), activity.Tags[0], activity.URL)
+		return
+	}
+
+	var exp *optimizev1beta2.Experiment
+	for i := range generatedResources {
+		if expObj, ok := generatedResources[i].(*optimizev1beta2.Experiment); ok {
+			exp = expObj
+			if exp.ObjectMeta.Annotations == nil {
+				exp.ObjectMeta.Annotations = make(map[string]string)
+			}
+
+			exp.ObjectMeta.Annotations[optimizev1beta2.AnnotationExperimentURL] = fmt.Sprintf("%s/%s", experimentURL, exp.ObjectMeta.Name)
```

Comment:
You can use the `metav1` helper here (it will allocate the missing map for you); also I think there is usually a bit of confusion over trailing slashes in URLs so it might be safer to account for both cases:

```go
metav1.SetMetaDataAnnotation(&exp.ObjectMeta, optimizev1beta2.AnnotationExperimentURL, strings.TrimRight(experimentURL, "/") + "/" + exp.Name)
```